### ISSUE TITLE
Reconsider allowing negative learning rates in torch.optim.SGD (and others)

### ISSUE THREAD

**Issue by latorrefabian** [pytorch/pytorch#81633]
### 🚀 The feature, motivation and pitch

Currently, optimizers throw an assertion error when negative learning rates are supplied at construction. This proposal suggests removing this restriction. This issue has been raised before in https://github.com/pytorch/pytorch/issues/46480. However the proposed solution (https://github.com/pytorch/pytorch/issues/68052) which will add a `maximize` parameter to the constructor is not satisfactory.

### Motivation
Since maximization is equivalent to minimizing the negative of a function, adding a `maximize` parameter to the SGD constructor does not add any new functionality but rather provides two ways of achieving the same thing, which contradicts the Python principle:

> There should be one– and preferably only one –obvious way to do it.*

On the other hand, Pytorch does not provide an easy (obvious) way to train models that are defined as a solution of a min-max problem. Perhaps the most popular case is the training of Generative Adversarial Networks, where the parameters of the *discriminator* and *generator* are the equilibrium points of a min-max game. Currently, in order train GANs with **Simultaneous Gradient Descent-Ascent** (or variants thereof) one has to create two optimizers and write two blocks of code that are almost the same, see for example the official pytorch tutorial https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html

### Alternatives

Instead, if we are allowed to set up a negative learning rate for some parameters, and positive learning rates for the rest of the parameters of a module, then **SGD naturally becomes (Stochastic) Gradient Descent Ascent**. Note that we are already allowed to set different learning rates for different parameter groups (unfortunately, they have to be both positive).

As a minimal working example consider the following:
```python
import torch

class GAN(torch.nn.Module):
    def __init__(
                self,
                generator: torch.nn.Module,
                discriminator: torch.nn.Module) -> None:
        self.generator = generator
        self.discriminator = discriminator

    def loss(self, real_data: torch.Tensor, noise: torch.Tensor) -> torch.Tensor:
        return self.discriminator(real_data) - self.discriminator(self.generator(noise))

gan = GAN(generator=torch.nn.Linear(1, 2), discriminator=torch.nn.Linear(2, 1))

gradient_descent_ascent = torch.optim.SGD([
                {'params': gan.generator.parameters(), 'lr': 1.},
                {'params': gan.discriminator.parameters(), 'lr': -1.}
            ], lr=1.)
```
In this case the parameters of the generator will try to minimize the objective function (positive learning rate), while the parameters of the discriminator will try to maximize the objective function (negative learning rate). With this change, it could be possible to train such models with the same structure of the training loop that is used in simpler tasks like classification:
```python
for real_data in train_loader:
    gradient_descent_ascent.zero_grad()
    noise = torch.randn(batch_size, 1)
    loss = gan.loss(real_data, noise)
    loss.backward()
    gradient_descent_ascent.step()
```
which is a much cleaner approach and would immediately provide the equivalent versions of Adam, Adagrad, etc. for min-max problems. Currently I am just copying the source code for SGD and removing the line that checks for positivity of the learning rate. Everything works as expected and the code is much cleaner when dealing with min-max problems. I cannot think of any risk in allowing negative learning rates. Throwing a warning might be enough, or adding a parameter ```allow_negative_lr``` with ```False``` as default might be more conservative (but clunky) solutions.

cc @vincentqb @jbschlosser @albanD

**Comment by latorrefabian** [comment]
@albanD what do you think?

**Comment by albanD** [comment]
Hey!
Thanks for taking the time to write down all of this.
The two main things I would like some clarification on are:
- Does the same happen when we use momentum or another optimizer than SGD?  SGD without momentum is a "weird" edge case where maximizing or flipping the lr is the same. But that is not always true. And it is not clear to me if what you want here is maximizing or flipping the lr?
- Following up on that, from the textual description you have, it seems to be that you do want to do "maximize". And so the same way you set different lr, you could set your two groups with lr=1.0 as global (as you already do) and one group with maximize=True and one with maximize=False. Doesn't that work for you?


> Since maximization is equivalent to minimizing the negative of a function, adding a maximize parameter to the SGD constructor does not add any new functionality but rather provides two ways of achieving the same thing, which contradicts the Python principle:

I don't think I agree with this statement. As your example shows, having such a flag allows a single optimizer to do maximization and minimization at the same time. While you would need two of them if you were flipping the function's value before computing the gradients.

**Comment by latorrefabian** [comment]
I see your point, and I think you are right, It should be possible to achieve the same effect by changing the `maximize` parameter for different groups. For some reason I thought it could only be set *globally* for the whole optimizer. I will try and see if it works.

Indeed I want the optimizer to update some parameters with minus (-) the gradient, and others with plus (-) the gradient. The purpose is to find a *saddle point* which is a solution of a min-max problem, that is, it is not a pure minimization or pure maximization problem.

I take back my statement if it is indeed possible to set a different `maximize` value per group as that effect cannot be achieved otherwise with a single optimizer object.

**Comment by latorrefabian** [comment]
I would really like to see an example where flipping the learning rate to negative does not achieve the exact same effect as `maximize=True`.

**Comment by albanD** [comment]
Here is a small toy example:
```python
import torch

def my_sgd(p, g, lr, mom, mom_buff):
    if mom != 0:
        mom_buff = mom * mom_buff + (1 - mom) * g
        update = mom_buff
    else:
        update = g
    return p - lr * g, mom_buff

inp = torch.rand(2)
g = torch.rand(2)
mom_buff = torch.rand(2)

print("No momentum")
print(my_sgd(inp, g, -1., 0., None))
print(my_sgd(inp, -g, 1., 0., None))


print("With momentum")
print(my_sgd(inp, g, -1., 0.5, mom_buff))
print(my_sgd(inp, -g, 1., 0.5, mom_buff))
```
This prints
```bash
$ python foo.py 
No momentum
(tensor([0.5049, 0.6166]), None)
(tensor([0.5049, 0.6166]), None)
With momentum
(tensor([0.5049, 0.6166]), tensor([0.3501, 0.3965]))
(tensor([0.5049, 0.6166]), tensor([-0.0344, -0.0367]))
```

As you can see, the one step is the same in this case but the updated momentum buffer is not (as expected as one see `g` while the other sees `-g`.
This means that the next step will use a different momentum buffer and thus start moving in a different direction.

**Comment by latorrefabian** [comment]
I think you are doing something wrong in your example, do you mean that after more iterations the tensors obtained with the two strategies would be different?

here is an example using the current SGD source code (change .txt to .py extension): I just copy-pasted the SGD source and

1. removed the constraint that the learning rate should be positive
2. removed all occurences of the `maximize` parameter
3. resolved some `if` statements involving such parameter, assuming `maximize=False`

Then the code just tracks two tensors that are initialized to the same value, and then updated for three iterations using the current SGD with `maximize=True` and momentum=0.5, or the proposed SGD (renamed here as CoolSGD) using negative learning rate and momentum=0.5. 

The result is that both tensors remain equal to each other.

[test_sgd_momentum.txt](https://github.com/pytorch/pytorch/files/9151994/test_sgd_momentum.txt)

**Comment by albanD** [comment]
Ho that's because sgd maximize is not correct :(  Fix here: https://github.com/pytorch/pytorch/pull/81859
You can mimick the correct behavior by doing these changes:
```python
    # When you create the optimizer
    normal_sgd = torch.optim.SGD([a,], lr=lr, maximize=False, momentum=0.5)

    # In the training loop, before the step
    a.grad = - a.grad
```

**Comment by albanD** [comment]
Note that sgd behavior was fixed and I think we have everything we want here right?

**Comment by latorrefabian** [comment]
yes all ok!, I can live with the `maximize` parameter although I believe allowing negative learning rates would not hurt.
